Include nltk data sets in nlp_primitives package install

dsherry opened the pull request:
Fix #28 

Changes:
* Add `nltk-data.tar` which contains the nltk corpuses we currently reference
* Add that data to the `nltk` data path at import-time
* Augment `sdist` to first unpack the data tarball, then include data files in the sdist via `MANIFEST.in`
* Add another set of CI jobs which ensure that we can make an nlp_primitives package, install that, and then still pass all the unit tests, vs an editable install
* Update `installdeps`/similar makefile targets to unpack the nltk data locally, for development env. Added [README section](https://github.com/FeatureLabs/nlp_primitives/tree/ds_package_data#development).
* Move some test utils into `test/`
* Add `make clean` command

To test manually: `python setup.py sdist && cd .. && pip install nlp_primitives/dist/nlp_primitives-1.0.0.tar.gz`, then check the `nltk.path`, and then try to run `nlp_primitives` code which relies on the `nltk` corpuses.

dsherry commented on setup.py:
```diff
@@ -10,6 +15,19 @@
 }
 
+class PreSDistCommand(sdist):
+    """Post-installation for installation mode."""
```
TODO change this

dsherry commented on setup.py:
```diff
@@ -10,6 +15,19 @@
+    """Post-installation for installation mode."""
+    def run(self):
+        nltk_data_tarball_path = pkg_resources.resource_filename('nlp_primitives', str(pathlib.Path('data','nltk-data.tar.gz')))
+        nltk_data_extract_path = pkg_resources.resource_filename('nlp_primitives', str(pathlib.Path('data')))
```
I'm still not sure on the pathing here, need to confirm this works

dsherry commented on .circleci/config.yml:
```diff
@@ -84,6 +84,54 @@ jobs:
+            NLP_PRIMITIVES_VERSION=$(python setup.py --version)
+            pip install dist/nlp_primitives-${NLP_PRIMITIVES_VERSION}.tar.gz
+      - run:
+          name: "Remove Repo Contents To Avoid Importing Source From Repo"
```
Ah, @rwedge  just pointed out that instead of messing with sys path to prevent the repo code from being imported in this test, we can just cd into a different directory. Will do

dsherry commented on dev-requirements.txt:
```diff
@@ -0,0 +1 @@
+wheel>=0.35.1
```
Needed to run `python setup.py bdist` which I did during some manual testing

dsherry commented on setup.py:
```diff
@@ -28,4 +52,7 @@
         ],
     },
+    cmdclass={
+        'sdist': PreSDistNLTKDataUnpackCommand
```
Its so cool that `setuptools` has an API for this!

dsherry commented on setup.py:
```diff
@@ -10,6 +15,23 @@
+        tar.extractall(path=nltk_data_extract_path)
+        tar.close()
+        print(f'Extraction of nltk data files complete')
+        sdist.run(self)
```
It would be great if someone could read this carefully. @rwedge @thehomebrewnerd @ctduffy @eccabay 

Based on my testing and reading of `pkg_resources`, I believe `pkg_resources` will always point to a directory in which the current user has readwrite access.

dsherry commented on MANIFEST.in:
```diff
@@ -1,5 +1,6 @@
 include *.txt
 include LICENSE
 include README.md
+recursive-include nlp_primitives/data/nltk-data/ *
```
This will include the data files once they've been un-tarred by the pre-sdist untarring command in `setup.py`

dsherry commented on .circleci/config.yml:
```diff
@@ -84,6 +84,50 @@ jobs:
+          command: |
+            source env/bin/activate
+            # change directory so that the repo code can't be imported, and we use the sdist installation instead
+            cd nlp_primitives
```
Thanks @rwedge for the tip here :)

eccabay commented on .circleci/config.yml:
```diff
@@ -84,6 +84,50 @@ jobs:
+      - run :
+          name: "Create Source Distribution"
+          command: |
+            virtualenv env -q
+            source env/bin/activate
```
Is it necessary to create this virtualenv here?

dsherry replied:
Oops, no! Doesn't do anything after the 1st time, I think, but I'll remove. Thanks

eccabay reviewed (changes requested):
Thanks for taking care of this! As one additional sanity check/since they won't be needed any more, could this PR also remove the downloads from the primitives that need nltk data?

rwedge commented on docs/source/changelog.rst:
```diff
@@ -1,6 +1,18 @@
+        * Add `nltk-data.tar.gz` which contains the nltk corpuses we currently reference :pr:`31`
+        * Add that data to the `nltk` data path at import-time :pr:`31`
+        * Augment `sdist` to first unpack the data tarball, then include data files in the sdist via `MANIFEST.in` :pr:`31`
+        * Add `make clean` command :pr:`31`
```
some of these changes are more developer-oriented than user-oriented, maybe we should add a "Development changes" section or something similar

dsherry replied:
@rwedge that's a good point. I'll try to clean this up a bit, and if it still feels weird, let's discuss again.

rwedge commented on docs/source/changelog.rst:
```diff
@@ -1,6 +1,18 @@
 =========
+**Future**
+    * Changes
+        * Add `nltk-data.tar.gz` which contains the nltk corpuses we currently reference :pr:`31`
+        * Add that data to the `nltk` data path at import-time :pr:`31`
```
I think we should make it clear that we're no longer downloading nltk data over the internet after installing the package

dsherry replied:
Yep thanks will do

dsherry commented on .circleci/config.yml:
```diff
@@ -114,4 +160,13 @@ workflows:
\ No newline at end of file
+          image_tag: "python:3.8"
+      - package_test:
+          name: "Packaging Test - Python 3.6"
```
That's a good question. The only benefit I could think of is that the non-sdist job type is simpler. If there's an error, maybe its easier to ssh into the box and poke around in the non-sdist job type. Beyond that, honestly not sure.

dsherry replied:
@rwedge there is a benefit after all to testing both packaged and non-packaged: the "non-packaged" test ensures the dev env works properly (sets up nltk data properly via makefile `installdeps`), whereas the "packaged" test ensures an sdist correctly includes and accesses all the nltk data.

dsherry commented on .circleci/config.yml:
```diff
@@ -84,6 +84,50 @@ jobs:
+          command: |
+            virtualenv env -q
+            source env/bin/activate
+            pip install -r requirements.txt
```
> install the package normally

Hmm, good point. `pip install nlp_primitives.tar.gz` should install deps. Will update

dsherry replied:
Hmm, its not necessary, but we do have to install the test requirements, and those install `complete-requirements.txt` under the hood. So I'm gonna leave this in.

dsherry replied:
Ah ok sure I'll try that.

dsherry commented:
@eccabay good point! Its weird to make this change without also switching off the downloads. Otherwise, the CI job we're adding won't actually fail if the data install isn't working correctly, since the download will still take place. I'll make that happen. Thanks!

dsherry commented on setup.py:
```diff
@@ -10,6 +15,23 @@
+        nltk_data_extract_path = pkg_resources.resource_filename('nlp_primitives', str(pathlib.Path('data')))
+        print(f'Extracting nltk data files from {nltk_data_tarball_path} to {nltk_data_extract_path}')
+        print(pkg_resources.resource_listdir('nlp_primitives', str(pathlib.Path('data'))))
+        tar = tarfile.open(nltk_data_tarball_path, "r:gz")
```
Note from Roy: try `with tar as tarfile.open(...):` to get scoped closing

eccabay reviewed (dismissed):
LGTM! Thanks for taking care of this.

dsherry commented:
@rwedge @eccabay some context on the changes I added:
* Removed all calls to `nltk.download()`, in favor of the nltk data tarball
* Updated our makefile so that `make-installdeps`/`make installdeps-complete`/`make installdeps-test` will unpack the nltk data locally, for developers. I had originally not realized this was necessary. Added dev install instructions to the README.md.
* Use a non-gzipped tarball, because the process of gzipping will *unzip* any `.zip` files included in the tarball, but we need `vader.zip` to stay sipped for nltk 🤦

rwedge commented on setup.py:
```diff
@@ -10,6 +15,23 @@
+        print(pkg_resources.resource_listdir('nlp_primitives', str(pathlib.Path('data'))))
+        with tarfile.open(nltk_data_tarball_path, "r") as tar:
+            tar.extractall(path=nltk_data_extract_path)
+            tar.close()
```
tar will close automatically when the with  block is done, we can remove this line and the code below can move outside the with block

dsherry commented:
@rwedge I addressed 2 of your 3 comments. For [this one](https://github.com/FeatureLabs/nlp_primitives/pull/31#discussion_r505513227), I tried to massage the order the deps are installed in, but I got [an error](https://app.circleci.com/pipelines/github/FeatureLabs/nlp_primitives/252/workflows/6a162765-4fbf-4695-97e7-9bb1f4c78c0e/jobs/2056). I suggest we merge what I have now.

dsherry commented:
ALso weird IDK why github dismissed Becca's review when I pushed 🤷‍♂️

rwedge reviewed (approved):
Looks good

dsherry commented:
Thanks!

PR coming soon to change the LSA primitive corpus. But otherwise, we should be ready for an `nlp_primitives` release imminently 🥳
